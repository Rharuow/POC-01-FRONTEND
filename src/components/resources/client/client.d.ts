import { Document } from "@/components/resources/document/document";
import { Address } from "@/components/resources/address/address";

export type Client = {
  id?: string;
  name?: string;
  email?: string;
  document?: Document;
  address?: Address;
};

export interface IFormClient {
  name: string;
  email: string;
  cpf: string;
  cnpj: string;
  billing: string;
  delivery: string;
}
