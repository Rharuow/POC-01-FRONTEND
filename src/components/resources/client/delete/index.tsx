import { Dialog, DialogContent, DialogFooter, DialogHeader, DialogTitle, DialogTrigger } from '@/components/ui/dialog'
import { Trash2 } from 'lucide-react'
import React, { useState } from 'react'
import { Client } from '../client'
import { Button } from '@/components/ui/button'
import { toast } from 'sonner'
import { useMutation } from '@apollo/client'
import { DELETE_CLIENT } from '@/service/mutation/delete/client'
import { GET_CLIENTS } from '@/service/queries/clients'
import { apolloClient } from '@/lib/apollo'

export default function DeleteClient({ client }: { client: Client }) {
  const [deleteModalIsOpen, setDeleteModalIsOpen] = useState(false)

  const [deleteOneClient] = useMutation(DELETE_CLIENT)

  async function handleDeleteClient(id: string) {
    try {
      await deleteOneClient({
        variables: { where: { id } },
        update: (cache, { data: { deleteOneClient } }) => {
          console.log("deleteOneClient = ", deleteOneClient);
          const { clients } = apolloClient.readQuery({ query: GET_CLIENTS });

          cache.writeQuery({
            query: GET_CLIENTS,
            data: {
              clients: clients.filter((client: Client) => client.id !== deleteOneClient.id),
            },
          });
        }
      })
      toast("Cliente deletado com sucesso...")
    } catch (error) {
      console.log("error = ", error);

      toast("Desculpe, mas não foi possível deletar o cliente")
    } finally {
      setDeleteModalIsOpen(false)
    }
  }

  return (
    <Dialog open={deleteModalIsOpen}>
      <DialogTrigger asChild onClick={() => setDeleteModalIsOpen(true)}>
        <Trash2 size={22} className="bg-red-500 p-1 text-white rounded-full" />
      </DialogTrigger>
      <DialogContent>
        <div className="flex flex-col gap-4">
          <DialogHeader>
            <DialogTitle><p>Você deseja deletar {client.name}?</p></DialogTitle>
          </DialogHeader>
          <DialogFooter>
            <div className="flex gap-2">
              <Button onClick={() => handleDeleteClient(String(client.id))}>Deletar</Button>
              <Button onClick={() => setDeleteModalIsOpen(false)}>Cancelar</Button>
            </div>
          </DialogFooter>
        </div>
      </DialogContent>
    </Dialog>
  )
}
