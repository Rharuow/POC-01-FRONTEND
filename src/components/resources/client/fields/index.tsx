import InputGroup from '@/components/ui/inputGroup';
import { Tabs, TabsContent, TabsList, TabsTrigger } from '@/components/ui/tabs';
import React, { useState } from 'react'
import { useFormContext } from 'react-hook-form';
import { IFormClient } from '../client';
import { cpfMask } from '@/lib/mask/cpf';
import { cpfIsValid } from '@/lib/validation/cpf';
import { cnpjIsValid } from '@/lib/validation/cnpj';
import { cnpjMask } from '@/lib/mask/cnpj';
import { Button } from '@/components/ui/button';
import Billing from './address/billing';
import Delivery from './address/delivery';

const Fields = ({ isLoading }: { isLoading: boolean }) => {

  const { setValue } = useFormContext<IFormClient>()

  const [typeDocument, setTypeDocument] = useState<"cpf" | "cnpj">("cpf");
  const [cpfIsInvalid, setCpfIsInvalid] = useState(false);
  const [cnpjIsInvalid, setCnpjIsInvalid] = useState(false);

  return (
    <>
      <div className="flex flex-col gap-2">
        <div className="flex flex-col gap-2 md:grid md:grid-cols-2">
          <InputGroup label="Nome" name="name" />
          <InputGroup label="Email" name="email" />
        </div>
        <Billing />
        <Delivery />
        <Tabs defaultValue={typeDocument}>
          <TabsList className="grid w-full grid-cols-2">
            <TabsTrigger
              value="cpf"
              onClick={() => {
                setCnpjIsInvalid(false);
                setTypeDocument("cpf");
              }}
            >
              CPF
            </TabsTrigger>
            <TabsTrigger
              value="cnpj"
              onClick={() => {
                setCpfIsInvalid(false);
                setTypeDocument("cnpj");
              }}
            >
              CNPJ
            </TabsTrigger>
          </TabsList>
          <TabsContent value="cpf">
            <InputGroup
              label="CPF"
              name="cpf"
              onChange={(event) => setValue("cpf", cpfMask(event.target.value))}
              onBlur={(event) => setCpfIsInvalid(!cpfIsValid(event.target.value))}
            />
          </TabsContent>
          <TabsContent value="cnpj">
            <InputGroup
              label="CNPJ"
              name="cnpj"
              onChange={(event) => {
                setCnpjIsInvalid(!cnpjIsValid(event.target.value));
                setValue("cnpj", cnpjMask(event.target.value));
              }}
              onBlur={(event) => setCnpjIsInvalid(!cnpjIsValid(event.target.value))}
            />
          </TabsContent>
        </Tabs>
      </div>
      <Button
        type="submit"
        disabled={
          cnpjIsInvalid ||
          cpfIsInvalid ||
          isLoading
        }
      >
        Salvar
      </Button>
    </>
  )
}

export default Fields
