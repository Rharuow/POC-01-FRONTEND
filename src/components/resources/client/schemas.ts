import { z } from "zod";

export const formClientSchema = z
  .object({
    name: z.string().min(4, { message: "Pelo menos 4 caracteres" }),
    cpf: z.string({ message: "CPF é obrigatório" }).optional(),
    cnpj: z.string({ message: "CNPJ é obrigatório" }).optional(),
    email: z
      .string({ required_error: "O email é obrigatório" })
      .email("Este é um email inválido."),
    billing: z.string({
      message: "Endereço de cobrança é obrigatório",
    }),
    delivery: z.string({ message: "Endereço de entrega é obrigatório" }),
    zipCode: z.string().optional(),
    neighborhood: z.string().optional(),
    state: z.string().optional(),
    city: z.string().optional(),
    number: z.string().optional(),
    street: z.string().optional(),
    complement: z.string().optional(),
  })
  .partial({
    cnpj: true,
    cpf: true,
  })
  .superRefine(({ cnpj, cpf, billing, delivery }, ctx) => {
    if (!cpf && !cnpj) {
      ctx.addIssue({
        code: z.ZodIssueCode.custom, // customize your issue
        message: "CPF é obrigatório",
        path: ["cpf"],
      });
      ctx.addIssue({
        code: z.ZodIssueCode.custom, // customize your issue
        message: "CNPJ é obrigatório",
        path: ["cnpj"],
      });
    }
    if (!billing) {
      ctx.addIssue({
        code: z.ZodIssueCode.custom,
        message: "Endereço de cobrança é obrigatório",
        path: ["billing"],
      });
    }
    if (!delivery) {
      ctx.addIssue({
        code: z.ZodIssueCode.custom,
        message: "Endereço de entrega é obrigatório",
        path: ["delivery"],
      });
    }

    return z.NEVER;
  });
