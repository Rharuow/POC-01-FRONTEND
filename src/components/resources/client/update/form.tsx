import React, { useState } from "react";
import { FormProvider, useForm } from "react-hook-form";
import { z } from "zod";
import { zodResolver } from "@hookform/resolvers/zod";
import { toast } from "sonner";

import { Button } from "@/components/ui/button";
import { Input } from "@/components/ui/input";
import { Tabs, TabsContent, TabsList, TabsTrigger } from "@/components/ui/tabs";
import { cpfMask } from "@/lib/mask/cpf";
import { cnpjMask } from "@/lib/mask/cnpj";
import { cpfIsValid } from "@/lib/validation/cpf";
import { cn } from "@/lib/utils";
import { cnpjIsValid } from "@/lib/validation/cnpj";
import { useMutation } from "@apollo/client";
import { Skeleton } from "@/components/ui/skeleton";
import { Client, IFormClient } from "../client";
import { UPDATE_CLIENT } from "@/service/mutation/client";
import { formClientSchema } from "../schemas";
import Loading from "./loading";
import InputGroup from "@/components/ui/inputGroup";

export const FormUpdateClient = ({ setEditModalIsOpen, client }: { setEditModalIsOpen: React.Dispatch<React.SetStateAction<boolean>>; client: Client }) => {
  const [updateOneClient] = useMutation(UPDATE_CLIENT);

  const methods = useForm<IFormClient>({
    resolver: zodResolver(formClientSchema),
    defaultValues: {
      billing: client.address?.billing,
      delivery: client.address?.delivery,
      cnpj: client.document?.cnpj,
      cpf: client.document?.cpf,
      email: client.email,
      name: client.name
    }
  });

  const [isLoading, setIsLoading] = useState(false);
  const [typeDocument, setTypeDocument] = useState<"cpf" | "cnpj">(client.document?.cnpj ? "cnpj" : "cpf");
  const [cpfIsInvalid, setCpfIsInvalid] = useState(false);
  const [cnpjIsInvalid, setCnpjIsInvalid] = useState(false);

  const {
    handleSubmit,
    register,
    setValue,
    formState: { errors },
  } = methods;

  async function onSubmit(data: IFormClient) {
    try {
      setIsLoading(true);

      toast("Cliente atualizado com sucesso");
      setEditModalIsOpen(false);
    } catch (error) {
      console.log(error);
      toast("Desculpe, algo está errado!");
    } finally {
      setIsLoading(false);
    }
  }

  return (
    <FormProvider {...methods}>
      <form onSubmit={handleSubmit(onSubmit)} className="flex flex-col gap-4">
        {
          isLoading ? <Loading /> : <div className="flex flex-col gap-2">
            <div className="flex flex-col gap-2 md:grid md:grid-cols-2">

              <InputGroup label="Nome" name="name" />

              <InputGroup label="Email" name="email" />

            </div>

            <InputGroup label="Endereço  de cobrança" name="billing" />

            <InputGroup label="Endreço de entrega" name="delivery" />

            <Tabs defaultValue={typeDocument}>
              <TabsList className="grid w-full grid-cols-2">
                <TabsTrigger
                  value="cpf"
                  disabled={isLoading}
                  onClick={() => {
                    setCnpjIsInvalid(false);
                    setTypeDocument("cpf");
                  }}
                >
                  CPF
                </TabsTrigger>
                <TabsTrigger
                  value="cnpj"
                  disabled={isLoading}
                  onClick={() => {
                    setCpfIsInvalid(false);
                    setTypeDocument("cnpj");
                  }}
                >
                  CNPJ
                </TabsTrigger>
              </TabsList>
              <TabsContent value="cpf">
                <InputGroup
                  label="CPF"
                  name="cpf"
                  onChange={(event) => setValue("cpf", cpfMask(event.target.value))}
                  onBlur={(event) => setCpfIsInvalid(!cpfIsValid(event.target.value))}
                />
              </TabsContent>
              <TabsContent value="cnpj">
                <InputGroup
                  label="CNPJ"
                  name="cnpj"
                  onChange={(event) => {
                    setCnpjIsInvalid(!cnpjIsValid(event.target.value));
                    setValue("cnpj", cnpjMask(event.target.value));
                  }}
                  onBlur={(event) => setCnpjIsInvalid(!cnpjIsValid(event.target.value))}
                />
              </TabsContent>
            </Tabs>
          </div>
        }

        <Button
          type="submit"
          disabled={
            Object.keys(errors).length !== 0 ||
            cnpjIsInvalid ||
            cpfIsInvalid ||
            isLoading
          }
        >
          Salvar
        </Button>
        <Button
          type="button"
          variant={"outline"}
          onClick={() => setEditModalIsOpen(false)}
        >
          Cancelar
        </Button>

      </form>
    </FormProvider>
  );
};
