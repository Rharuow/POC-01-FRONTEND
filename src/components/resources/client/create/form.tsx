import React, { useState } from "react";
import { FormProvider, useForm } from "react-hook-form";
import { z } from "zod";
import { zodResolver } from "@hookform/resolvers/zod";
import { toast } from "sonner";

import { Button } from "@/components/ui/button";
import { Input } from "@/components/ui/input";
import { Tabs, TabsContent, TabsList, TabsTrigger } from "@/components/ui/tabs";
import { cpfMask } from "@/lib/mask/cpf";
import { cnpjMask } from "@/lib/mask/cnpj";
import { cpfIsValid } from "@/lib/validation/cpf";
import { cn } from "@/lib/utils";
import { cnpjIsValid } from "@/lib/validation/cnpj";
import { useMutation } from "@apollo/client";
import { Skeleton } from "@/components/ui/skeleton";
import { useOpenCreateClientModalContext } from "../list";
import { GET_CLIENTS } from "@/service/queries/clients";
import { apolloClient } from "@/lib/apollo";
import { CREATE_CLIENT } from "@/service/mutation/client";

interface IFormCreateClient {
  name: string;
  email: string;
  documentation: string;
  billing: string;
  delivery: string;
}

type ClientCreateInput = {
  data: {
    address: {
      connectOrCreate: {
        create: {
          billing: string;
          delivery: string;
        };
        where: {
          billing: string;
          delivery: string;
        };
      }
    };
    document: {
      connectOrCreate: {
        create: {
          cnpj?: string;
          cpf?: string;
        };
        where: {
          cnpj?: string;
          cpf?: string;
        };
      }
    };
    email: string;
    name: string;
  };
};

const schema = z.object({
  name: z.string().min(4, { message: "Pelo menos 4 caracteres" }),
  documentation: z.string().min(1, { message: "Documento inválido" }),
  email: z
    .string({ required_error: "O email é obrigatório" })
    .email("Este é um email inválido."),
  billing: z.string().min(1, { message: "Endereço de cobrança inválido" }),
  delivery: z.string().min(1, { message: "Endereço de entrega inválido" }),
});

export const FormCreateClient = () => {
  const [createOneClient] = useMutation(CREATE_CLIENT);
  const { setIsOpen } = useOpenCreateClientModalContext();

  const methods = useForm<IFormCreateClient>({
    resolver: zodResolver(schema),
  });

  const [isLoading, setIsLoading] = useState(false);
  const [typeDocument, setTypeDocument] = useState<"cpf" | "cnpj">("cpf");
  const [cpfIsInvalid, setCpfIsInvalid] = useState(false);
  const [cnpjIsInvalid, setCnpjIsInvalid] = useState(false);

  const {
    handleSubmit,
    register,
    setValue,
    formState: { errors },
  } = methods;

  async function onSubmit(data: IFormCreateClient) {
    try {
      setIsLoading(true);
      const { documentation, billing, delivery, ...dataSpread } = data;
      const formattedData: ClientCreateInput = {
        data: {
          ...dataSpread,
          document: {
            connectOrCreate: {
              create: {
                ...(typeDocument === "cnpj"
                  ? { cnpj: documentation }
                  : { cpf: documentation }),
              },
              where: {
                ...(typeDocument === "cnpj"
                  ? { cnpj: documentation }
                  : { cpf: documentation }),
              }
            },
          },
          address: {
            connectOrCreate: {
              create: {
                billing,
                delivery,
              },
              where: {
                billing,
                delivery,
              }
            },
          },
        },
      };

      await createOneClient({
        variables: formattedData,
        update: (cache, { data: { createOneClient } }) => {
          const { clients } = apolloClient.readQuery({ query: GET_CLIENTS });

          cache.writeQuery({
            query: GET_CLIENTS,
            data: {
              clients: [...clients, createOneClient],
            },
          });
        },
      });
      toast("Cliente criado com sucesso");
      setIsOpen(false);
    } catch (error) {
      console.log(error);
      toast("Desculpe, algo está errado!");
    } finally {
      setIsLoading(false);
    }
  }

  return (
    <FormProvider {...methods}>
      <form onSubmit={handleSubmit(onSubmit)} className="flex flex-col gap-2">
        <div className="flex flex-col gap-2 md:grid md:grid-cols-2">
          {isLoading ? (
            <Skeleton className="w-full h-10 rounded-full" />
          ) : (
            <div className="flex flex-col">
              <Input
                label="Nome"
                {...register("name")}
                autoFocus={false}
                className={cn({
                  "border border-red-700": errors && errors.name,
                })}
              />
              {errors && errors.name && (
                <span className="text-xs text-red-400 font-bold">
                  {errors.name.message}
                </span>
              )}
            </div>
          )}
          {isLoading ? (
            <Skeleton className="w-full h-10 rounded-full" />
          ) : (
            <div className="flex flex-col">
              <Input
                label="Email"
                {...register("email")}
                className={cn({
                  "border border-red-700": errors && errors.email,
                })}
              />
              {errors && errors.email && (
                <span className="text-xs text-red-400 font-bold">
                  {errors.email.message}
                </span>
              )}
            </div>
          )}
        </div>
        {isLoading ? (
          <Skeleton className="w-full h-10 rounded-full" />
        ) : (
          <div className="flex flex-col">
            <Input
              label="Endreço de cobrança"
              {...register("billing")}
              className={cn({
                "border border-red-700": errors && errors.billing,
              })}
            />
            {errors && errors.billing && (
              <span className="text-xs text-red-400 font-bold">
                {errors.billing.message}
              </span>
            )}
          </div>
        )}
        {isLoading ? (
          <Skeleton className="w-full h-10 rounded-full" />
        ) : (
          <div className="flex flex-col">
            <Input
              label="Endreço de entrega"
              {...register("delivery")}
              className={cn({
                "border border-red-700": errors && errors.delivery,
              })}
            />
            {errors && errors.delivery && (
              <span className="text-xs text-red-400 font-bold">
                {errors.delivery.message}
              </span>
            )}
          </div>
        )}

        <Tabs defaultValue={typeDocument}>
          <TabsList className="grid w-full grid-cols-2">
            <TabsTrigger
              value="cpf"
              disabled={isLoading}
              onClick={() => {
                setValue("documentation", "");
                setCnpjIsInvalid(false);
                setTypeDocument("cpf");
              }}
            >
              CPF
            </TabsTrigger>
            <TabsTrigger
              value="cnpj"
              disabled={isLoading}
              onClick={() => {
                setValue("documentation", "");
                setCpfIsInvalid(false);
                setTypeDocument("cnpj");
              }}
            >
              CNPJ
            </TabsTrigger>
          </TabsList>
          <TabsContent value="cpf">
            {isLoading ? (
              <Skeleton className="w-full h-10 rounded-full" />
            ) : (
              <div className="flex flex-col">
                <Input
                  label="CPF"
                  inputMode="numeric"
                  {...register("documentation", {
                    onChange: (event) => {
                      setValue("documentation", cpfMask(event.target.value));
                    },
                    onBlur: (event) => {
                      setCpfIsInvalid(!cpfIsValid(event.target.value));
                    },
                  })}
                  className={cn({
                    "border border-red-700":
                      (errors && errors.documentation) || cpfIsInvalid,
                  })}
                />
                {((errors && errors.documentation) || cpfIsInvalid) && (
                  <span className="text-xs text-red-400 font-bold">
                    {errors.documentation?.message || "Formato do CPF inválido"}
                  </span>
                )}
              </div>
            )}
          </TabsContent>
          <TabsContent value="cnpj">
            {isLoading ? (
              <Skeleton className="w-full h-10 rounded-full" />
            ) : (
              <div className="flex flex-col">
                <Input
                  label="CNPJ"
                  inputMode="numeric"
                  {...register("documentation", {
                    onChange: (event) => {
                      setCnpjIsInvalid(!cnpjIsValid(event.target.value));
                      setValue("documentation", cnpjMask(event.target.value));
                    },
                    onBlur: (event) => {
                      console.log(cnpjIsValid(event.target.value));
                      setCnpjIsInvalid(!cnpjIsValid(event.target.value));
                    },
                  })}
                  className={cn({
                    "border border-red-700":
                      (errors && errors.documentation) || cnpjIsInvalid,
                  })}
                />
                {((errors && errors.documentation) || cnpjIsInvalid) && (
                  <span className="text-xs text-red-400 font-bold">
                    {errors.documentation?.message ||
                      "Formato do CNPJ inválido"}
                  </span>
                )}
              </div>
            )}
          </TabsContent>
        </Tabs>

        <Button
          type="submit"
          disabled={
            Object.keys(errors).length !== 0 ||
            cnpjIsInvalid ||
            cpfIsInvalid ||
            isLoading
          }
        >
          Salvar
        </Button>
      </form>
    </FormProvider>
  );
};
