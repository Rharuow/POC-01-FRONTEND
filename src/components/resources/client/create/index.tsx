import { Button } from "@/components/ui/button";
import {
  DialogContent,
  DialogFooter,
  DialogHeader,
  DialogTitle,
} from "@/components/ui/dialog";
import React from "react";
import { FormCreateClient } from "./form";

export const CreateClient = () => {
  return (
    <DialogContent className="max-w-[425px] md:max-w-[525px] lg:max-w-[825px]">
      <DialogHeader>
        <DialogTitle>Adicionar Cliente</DialogTitle>
      </DialogHeader>
      <FormCreateClient />
      <DialogFooter>
        <Button type="submit">Salvar</Button>
      </DialogFooter>
    </DialogContent>
  );
};
