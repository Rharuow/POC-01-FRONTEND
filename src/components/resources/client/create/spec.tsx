import "@testing-library/jest-dom";
import { screen, render, fireEvent } from '@testing-library/react'

import { MockedProvider } from "@apollo/client/testing";
import { FormCreateClient } from "./form";
import { Dialog } from "@/components/ui/dialog";

const setup = () => {
  const utils = render(
    <MockedProvider mocks={[]} addTypename={false}>
      <Dialog open={true}>
        <FormCreateClient />
      </Dialog>
    </MockedProvider>
  )

  const nameField = screen.getByLabelText('Nome')
  const btnSubmit = screen.getByText('Salvar')

  return {
    nameField,
    btnSubmit,
    ...utils
  }
}

it('Verificando renderização do component de formulário de criação de cliente', () => {
  const { container } = setup()

  expect(container).toMatchSnapshot()
})

it('Verificando se todos os campos são obrigatórios, apresentando o border-red-700 no attributo class', async () => {
  const { btnSubmit } = setup()

  fireEvent.submit(btnSubmit)

  expect((await screen.findAllByRole("textbox")).filter(field => field.className.includes("border-red-700"))).toHaveLength(5)
})