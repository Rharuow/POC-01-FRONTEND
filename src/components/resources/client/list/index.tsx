import React, { createContext, useContext, useState } from "react";
import { PlusCircle } from "lucide-react";

import {
  Carousel,
  CarouselContent,
  CarouselItem,
} from "@/components/ui/carousel";
import { Skeleton } from "@/components/ui/skeleton";
import { GET_CLIENTS } from "@/service/queries/clients";
import { useQuery } from "@apollo/client";
import { CardClient } from "./card";
import { cn } from "@/lib/utils";
import { Card, CardContent } from "@/components/ui/card";
import { Dialog, DialogClose, DialogTrigger } from "@/components/ui/dialog";
import { CreateClient } from "../create";
import { Client } from "../client";

const OpenCreateClientModalContext = createContext<{
  isOpen: boolean;
  setIsOpen: React.Dispatch<React.SetStateAction<boolean>>;
}>({
  isOpen: false,
  setIsOpen: () => {},
});

export const useOpenCreateClientModalContext = () =>
  useContext(OpenCreateClientModalContext);

export const ListClient = () => {
  const { data, loading } = useQuery<{ clients: Array<Client> }>(GET_CLIENTS);

  const [isOpenCreateClientModal, setIsOpenCreateClientModal] = useState(false);

  return (
    <Carousel className="w-full">
      <CarouselContent className="w-full">
        {loading
          ? Array.from({ length: 4 }).map((_, index) => (
              <CarouselItem
                className="md:basis-1/2 lg:basis-1/4 flex justify-center"
                key={index}
              >
                <Skeleton className="w-full h-20" />
              </CarouselItem>
            ))
          : data?.clients.map((client, _, self) => (
              <CarouselItem
                className={cn("md:basis-1/2", {
                  "lg:basis-1/4": self.length >= 4,
                  "lg:basis-1/3": self.length === 3,
                  "lg:basis-1/2": self.length <= 2,
                })}
                key={client.id}
              >
                <CardClient client={client} />
              </CarouselItem>
            ))}
        <CarouselItem
          className={cn("md:basis-1/2", {
            "lg:basis-1/4": self.length >= 4,
            "lg:basis-1/3": self.length === 3,
            "lg:basis-1/2": self.length <= 2,
            "lg:basis-auto flex justify-center grow": self.length === 0,
          })}
          key={data?.clients.length}
        >
          <OpenCreateClientModalContext.Provider
            value={{
              isOpen: isOpenCreateClientModal,
              setIsOpen: setIsOpenCreateClientModal,
            }}
          >
            <Dialog open={isOpenCreateClientModal}>
              <DialogTrigger
                asChild
                onClick={() => setIsOpenCreateClientModal(true)}
              >
                <Card
                  className={cn(
                    "h-full w-full bg-transparent text-white border-dashed hover:cursor-pointer p-0",
                    {
                      "min-h-20": self.length === 0,
                    }
                  )}
                >
                  <CardContent className="flex flex-col gap-4 justify-center items-center p-4">
                    <PlusCircle />
                    {self.length === 0 ? (
                      <p>Nenhum cliente Cadastrado</p>
                    ) : (
                      <p>Adicionar Cliente</p>
                    )}
                  </CardContent>
                </Card>
              </DialogTrigger>
              <CreateClient />
            </Dialog>
          </OpenCreateClientModalContext.Provider>
        </CarouselItem>
      </CarouselContent>
    </Carousel>
  );
};
