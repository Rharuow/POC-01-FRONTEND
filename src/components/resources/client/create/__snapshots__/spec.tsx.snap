// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Verificando renderização do component de formulário de criação de cliente 1`] = `
<div>
  <form
    class="flex flex-col gap-4"
  >
    <div
      class="flex flex-col gap-2"
    >
      <div
        class="flex flex-col gap-2 md:grid md:grid-cols-2"
      >
        <div
          class="flex flex-col"
        >
          <div
            class="relative"
          >
            <input
              class="flex h-10 w-full border z-10 border-input rounded-full bg-transparent px-3 py-2 text-sm file:border-0 file:bg-transparent file:text-sm file:font-medium placeholder:text-muted-foreground focus-visible:outline-none disabled:cursor-not-allowed disabled:opacity-50"
              id="name"
              name="name"
            />
            <label
              class="text-gray-500 absolute top-1/2 -translate-y-1/2 left-3 animate-label-blur"
              for="name"
            >
              Nome
            </label>
          </div>
        </div>
        <div
          class="flex flex-col"
        >
          <div
            class="relative"
          >
            <input
              class="flex h-10 w-full border z-10 border-input rounded-full bg-transparent px-3 py-2 text-sm file:border-0 file:bg-transparent file:text-sm file:font-medium placeholder:text-muted-foreground focus-visible:outline-none disabled:cursor-not-allowed disabled:opacity-50"
              id="email"
              name="email"
            />
            <label
              class="text-gray-500 absolute top-1/2 -translate-y-1/2 left-3 animate-label-blur"
              for="email"
            >
              Email
            </label>
          </div>
        </div>
      </div>
      <button
        aria-controls="radix-:r3:"
        aria-expanded="false"
        aria-haspopup="dialog"
        data-state="closed"
        type="button"
      >
        <div
          class="flex flex-col"
        >
          <div
            class="relative"
          >
            <input
              class="flex h-10 w-full border z-10 border-input rounded-full bg-transparent px-3 py-2 text-sm file:border-0 file:bg-transparent file:text-sm file:font-medium placeholder:text-muted-foreground focus-visible:outline-none disabled:cursor-not-allowed disabled:opacity-50"
              id="billing"
              name="billing"
              readonly=""
            />
            <label
              class="text-gray-500 absolute top-1/2 -translate-y-1/2 left-3 animate-label-blur"
              for="billing"
            >
              Endereço  de cobrança
            </label>
          </div>
        </div>
      </button>
      <button
        aria-controls="radix-:r4:"
        aria-expanded="false"
        aria-haspopup="dialog"
        data-state="closed"
        type="button"
      >
        <div
          class="flex flex-col"
        >
          <div
            class="relative"
          >
            <input
              class="flex h-10 w-full border z-10 border-input rounded-full bg-transparent px-3 py-2 text-sm file:border-0 file:bg-transparent file:text-sm file:font-medium placeholder:text-muted-foreground focus-visible:outline-none disabled:cursor-not-allowed disabled:opacity-50"
              id="delivery"
              name="delivery"
              readonly=""
            />
            <label
              class="text-gray-500 absolute top-1/2 -translate-y-1/2 left-3 animate-label-blur"
              for="delivery"
            >
              Endreço de entrega
            </label>
          </div>
        </div>
      </button>
      <div
        data-orientation="horizontal"
        dir="ltr"
      >
        <div
          aria-orientation="horizontal"
          class="h-10 items-center justify-center rounded-md bg-muted p-1 text-muted-foreground grid w-full grid-cols-2"
          data-orientation="horizontal"
          role="tablist"
          style="outline: none;"
          tabindex="0"
        >
          <button
            aria-controls="radix-:r5:-content-cpf"
            aria-selected="true"
            class="inline-flex items-center justify-center whitespace-nowrap rounded-sm px-3 py-1.5 text-sm font-medium ring-offset-background transition-all focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-ring focus-visible:ring-offset-2 disabled:pointer-events-none disabled:opacity-50 data-[state=active]:bg-background data-[state=active]:text-foreground data-[state=active]:shadow-sm"
            data-orientation="horizontal"
            data-radix-collection-item=""
            data-state="active"
            id="radix-:r5:-trigger-cpf"
            role="tab"
            tabindex="-1"
            type="button"
          >
            CPF
          </button>
          <button
            aria-controls="radix-:r5:-content-cnpj"
            aria-selected="false"
            class="inline-flex items-center justify-center whitespace-nowrap rounded-sm px-3 py-1.5 text-sm font-medium ring-offset-background transition-all focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-ring focus-visible:ring-offset-2 disabled:pointer-events-none disabled:opacity-50 data-[state=active]:bg-background data-[state=active]:text-foreground data-[state=active]:shadow-sm"
            data-orientation="horizontal"
            data-radix-collection-item=""
            data-state="inactive"
            id="radix-:r5:-trigger-cnpj"
            role="tab"
            tabindex="-1"
            type="button"
          >
            CNPJ
          </button>
        </div>
        <div
          aria-labelledby="radix-:r5:-trigger-cpf"
          class="mt-2 ring-offset-background focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-ring focus-visible:ring-offset-2"
          data-orientation="horizontal"
          data-state="active"
          id="radix-:r5:-content-cpf"
          role="tabpanel"
          style="animation-duration: 0s;"
          tabindex="0"
        >
          <div
            class="flex flex-col"
          >
            <div
              class="relative"
            >
              <input
                class="flex h-10 w-full border z-10 border-input rounded-full bg-transparent px-3 py-2 text-sm file:border-0 file:bg-transparent file:text-sm file:font-medium placeholder:text-muted-foreground focus-visible:outline-none disabled:cursor-not-allowed disabled:opacity-50"
                id="cpf"
                name="cpf"
              />
              <label
                class="text-gray-500 absolute top-1/2 -translate-y-1/2 left-3 animate-label-blur"
                for="cpf"
              >
                CPF
              </label>
            </div>
          </div>
        </div>
        <div
          aria-labelledby="radix-:r5:-trigger-cnpj"
          class="mt-2 ring-offset-background focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-ring focus-visible:ring-offset-2"
          data-orientation="horizontal"
          data-state="inactive"
          hidden=""
          id="radix-:r5:-content-cnpj"
          role="tabpanel"
          tabindex="0"
        />
      </div>
    </div>
    <button
      class="inline-flex items-center justify-center whitespace-nowrap rounded-md text-sm font-medium ring-offset-background transition-colors focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-ring focus-visible:ring-offset-2 disabled:pointer-events-none disabled:opacity-50 bg-gray-900 text-white hover:bg-gray-900/90 h-10 px-4 py-2"
      type="submit"
    >
      Salvar
    </button>
  </form>
</div>
`;
